orca/producer.go

```diff
@@ -0,0 +1,219 @@
+	listeners map[OOBListener]struct{} // set of registered listeners
+}
+
+// registerListener adds the listener and its requested report interval to the
+// producer.
+func (p *producer) registerListener(l OOBListener, interval time.Duration) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.listeners[l] = struct{}{}
+	p.intervals[interval]++
+}
+
+// registerListener removes the listener and its requested report interval to
+// the producer.
+func (p *producer) unregisterListener(l OOBListener, interval time.Duration) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	delete(p.listeners, l)
+	p.intervals[interval]--
+	if p.intervals[interval] == 0 {
+		delete(p.intervals, interval)
+	}
+}
+
+// minInterval returns the smallest key in p.intervals.
+func (p *producer) minInterval() time.Duration {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	min := time.Duration(-1)
+	for t := range p.intervals {
+		if min == time.Duration(-1) || t < min {
+			min = t
+		}
+	}
+	return min
+}
+
+// run manages the ORCA OOB stream on the subchannel.
+func (p *producer) run(ctx context.Context) {
+	defer p.closed.Fire()
+	backoffAttempt := 0
+	backoffTimer := time.NewTimer(0)
+	for ctx.Err() == nil {
+		select {
+		case <-backoffTimer.C:
+		case <-ctx.Done():
+			return
+		}
+
+		resetBackoff, err := p.runStream(ctx)
+
+		if resetBackoff {
+			backoffTimer.Reset(0)
+			backoffAttempt = 0
+		} else {
+			backoffTimer.Reset(p.backoff(backoffAttempt))
+			backoffAttempt++
+		}
+
+		switch {
+		case err == nil:
+			// No error was encountered; restart the stream.
+		case ctx.Err() != nil:
+			// Producer was stopped; exit immediately and without logging an
+			// error.
+			return
+		case status.Code(err) == codes.Unimplemented:
+			// Unimplemented; do not retry and log at ERROR level.
+			logger.Error("Server doesn't support ORCA OOB load reporting protocol; not listening for load reports.")
+			return
+		case status.Code(err) == codes.Unavailable:
+			// The SubConn is not currently ready; backoff silently.
+			//
+			// TODO: don't attempt the stream until the state is READY to
+			// minimize the chances of this case and to avoid using the
+			// exponential backoff mechanism, as we should know it's safe to
+			// retry when the state is READY again.
+		default:
+			// Log all other stream errors.
+			logger.Info("Received unexpected stream error:", err, ctx.Err())
+		}
+	}
+}
+
+// runStream runs a single stream on the subchannel and returns the resulting
+// error, if any, and whether or not the run loop should reset the backoff
+// timer to zero or advance it.
+func (p *producer) runStream(ctx context.Context) (resetBackoff bool, err error) {
+	interval := p.minInterval()
+	streamCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	stream, err := p.client.StreamCoreMetrics(streamCtx, &v3orcaservicepb.OrcaLoadReportRequest{
+		ReportInterval: durationpb.New(interval),
+	})
+	if err != nil {
+		return resetBackoff, err
```

Comment:
Can we do `return false, err` here instead, like how we do at the end of this function?

Also, why are named return values useful in this function?

Reply:
> Can we do return false, err here instead, like how we do at the end of this function?

We could.  I don't feel strongly about this and flipped it late during implementation.

> Also, why are named return values useful in this function?

Documentation purposes mostly.